### PR TITLE
Implement a node iterator

### PR DESCRIPTION
**Pull request by gsalgado**
My plan is to use this to validate the state trie we download after a fast-sync.
I've looked at pyethereum for "inspiration" (i.e. lifted most of the code from
there), and have been trying to make it more straightforward/cleaner, but I'm
kinda stuck, so decided to submit it as is

### CODE REVIEW

File: tests/test_iter.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+def test_trie_next_prev_using_fixtures(fixture_name, fixture):
+    trie = HexaryTrie(db={})
+    for k in fixture['in']:
+        k = force_bytes(k)
+        trie[k] = k
+
+    iterator = NodeIterator(trie)
+    for point, _, nxt in fixture['tests']:
+        point = force_bytes(point)
+        nxt = force_bytes(nxt)
```
Comment (pipermerriam) [inline tests/test_iter.py]: You shouldn't use `force_bytes` anymore.  It's going to be deprecated due to some unfixable encoding issues.

File: trie/iter.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            return curr_key
+        elif node_type == NODE_TYPE_EXTENSION:
+            curr_key = extract_key(node)
+            sub_node = self.trie._get_node(node[1])
+            return curr_key + self._get_next(sub_node)
+        elif node_type == NODE_TYPE_BRANCH:
+            if node[16]:
+                return (16,)
+            scan_range = range(16)
+            for i in scan_range:
```
Comment (pipermerriam) [inline trie/iter.py]: *nitpick*

Odd to me that this isn't just `for i in range(16)`

File: trie/iter.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    def _get_next(self, node):
+        node_type = get_node_type(node)
+        if node_type == NODE_TYPE_BLANK:
+            return None
+        elif node_type == NODE_TYPE_LEAF:
+            curr_key = extract_key(node)
+            return curr_key
+        elif node_type == NODE_TYPE_EXTENSION:
+            curr_key = extract_key(node)
+            sub_node = self.trie._get_node(node[1])
```
Comment (pipermerriam) [inline trie/iter.py]: It'd be nice if we could avoid using the internal `trie._get_node`.  Maybe instead we could add a new method on the trie class for retrieving *raw* nodes or something?  `Trie.get_raw_node`?

**Review by pipermerriam — APPROVED** [review]
Looks good.  We can address performance issues later when we actually know what they are.

File: trie/iter.py (hunk 2 above)
Comment (gsalgado) [reply trie/iter.py]: Yep, that was some leftover from a refactoring, will change

File: trie/iter.py (hunk 3 above)
Comment (gsalgado) [reply trie/iter.py]: Well, isn't that just a matter of renaming it to `trie.get_node()` (I'd prefer that over `get_raw_node()` as the latter kind of suggests it'd return the RLP-encoded node)?

File: tests/test_iter.py (hunk 1 above)
Comment (gsalgado) [reply tests/test_iter.py]: Should I just `.encode('utf-8')` until the new APIs land?

Comment (pipermerriam) [reply tests/test_iter.py]: Yes, I think that is fine.

File: trie/iter.py (hunk 3 above)
Comment (pipermerriam) [reply trie/iter.py]: My mistake.  `trie.get_node()` is fine.  For some reason I thought there was already a `get_node`, hence the new name suggestion.

**Pull request merged**